pkg/logql/probabilistic_evaluator.go

```diff
@@ -0,0 +1,444 @@
+	"fmt"
+	"math"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql"
+	promql_parser "github.com/prometheus/prometheus/promql/parser"
+
+	"github.com/grafana/dskit/tenant"
+
+	"github.com/grafana/loki/pkg/logql/sketch"
+	"github.com/grafana/loki/pkg/logql/syntax"
+	"github.com/grafana/loki/pkg/logqlmodel"
+	"github.com/grafana/loki/pkg/util"
+	"github.com/grafana/loki/pkg/util/validation"
+)
+
+type T int64
+
+const (
+	VecType T = iota
+	TopKVecType
+)
+
+type StepResult interface {
+	Type() T
+
+	SampleVector() promql.Vector
+	TopkVector() sketch.TopKVector
+}
+
+type SampleVector promql.Vector
+
+func (SampleVector) Type() T {
+	return VecType
+}
+
+func (s SampleVector) SampleVector() promql.Vector {
+	return promql.Vector(s)
+}
+
+func (s SampleVector) TopkVector() sketch.TopKVector {
+	return sketch.TopKVector{}
+}
+
+type TopKVector sketch.TopKVector
+
+func (TopKVector) Type() T {
+	return TopKVecType
+}
+
+func (v TopKVector) SampleVector() promql.Vector {
+	return nil
+}
+
+func (v TopKVector) TopkVector() sketch.TopKVector {
+	return sketch.TopKVector(v)
+}
+
+// ProbabilisticStepEvaluator evaluate a single step of a probabilistic query type.
+type ProbabilisticStepEvaluator interface {
+	// while Next returns a promql.Value, the only acceptable types are Scalar and Vector.
+	Next() (ok bool, ts int64, r StepResult)
+	// Close all resources used.
+	Close() error
+	// Reports any error
+	Error() error
+
+	Type() T
+}
+
+type ProbabilisticEvaluator struct {
+	Evaluator // this evaluator should be a DefaultEvaluator
+	logger    log.Logger
+}
+
+type pStepEvaluator struct {
+	fn    func() (bool, int64, StepResult)
+	close func() error
+	err   func() error
+	t     T
+}
+
+func (e *pStepEvaluator) Type() T {
+	return e.t
+}
+
+func (e *pStepEvaluator) Next() (bool, int64, StepResult) {
+	return e.fn()
+}
+
+func (e *pStepEvaluator) Close() error {
+	return e.close()
+}
+
+func (e *pStepEvaluator) Error() error {
+	return e.err()
+}
+
+// this function is just copy paste from the regular step eval, maybe we don't need it?
+func NewProbabilisticStepEvaluator(fn func() (bool, int64, StepResult), closeFn func() error, err func() error) (ProbabilisticStepEvaluator, error) {
+	if fn == nil {
+		return nil, errors.New("nil step evaluator fn")
+	}
+
+	if closeFn == nil {
+		closeFn = func() error { return nil }
+	}
+
+	if err == nil {
+		err = func() error { return nil }
+	}
+	return &pStepEvaluator{
+		fn:    fn,
+		close: closeFn,
+		err:   err,
+	}, nil
+}
+
+// NewDefaultEvaluator constructs a DefaultEvaluator
+func NewProbabilisticEvaluator(querier Querier, maxLookBackPeriod time.Duration) Evaluator {
+	d := NewDefaultEvaluator(querier, maxLookBackPeriod)
+	p := &ProbabilisticEvaluator{Evaluator: d}
+	return p
+}
+
+type StepEvaluatorAdapter struct {
+	d StepEvaluator
+}
```

Comment:
did I miss something, seems like we're only using this type in this package?

Reply:
I've used it original for the merging. Will revert it now.